AK/Optional.h

```diff
@@ -95,42 +95,42 @@ requires(!IsLvalueReference<T>) class [[nodiscard]] Optional<T> {
 
     ALWAYS_INLINE Optional(Optional const& other)
     requires(!IsTriviallyCopyConstructible<T>)
-        : m_has_value(other.m_has_value)
+        : m_null(), m_has_value(other.m_has_value)
     {
         if (other.has_value())
-            new (&m_storage) T(other.value());
+            construct_at(&m_storage, other.value());
     }
 
     ALWAYS_INLINE Optional(Optional&& other)
-        : m_has_value(other.m_has_value)
+        : m_null(), m_has_value(other.m_has_value)
     {
         if (other.has_value())
-            new (&m_storage) T(other.release_value());
+            construct_at(&m_storage, other.release_value());
     }
 
     template<typename U>
     requires(IsConstructible<T, U const&> && !IsSpecializationOf<T, Optional> && !IsSpecializationOf<U, Optional>) ALWAYS_INLINE explicit Optional(Optional<U> const& other)
-        : m_has_value(other.m_has_value)
+        : m_null(), m_has_value(other.m_has_value)
     {
         if (other.has_value())
-            new (&m_storage) T(other.value());
+            construct_at(&m_storage, other.value());
     }
 
     template<typename U>
     requires(IsConstructible<T, U &&> && !IsSpecializationOf<T, Optional> && !IsSpecializationOf<U, Optional>) ALWAYS_INLINE explicit Optional(Optional<U>&& other)
-        : m_has_value(other.m_has_value)
+        : m_null(), m_has_value(other.m_has_value)
     {
         if (other.has_value())
-            new (&m_storage) T(other.release_value());
+            construct_at(&m_storage, other.release_value());
     }
 
     template<typename U = T>
     requires(!IsSame<OptionalNone, RemoveCVReference<U>>)
     ALWAYS_INLINE explicit(!IsConvertible<U&&, T>) Optional(U&& value)
     requires(!IsSame<RemoveCVReference<U>, Optional<T>> && IsConstructible<T, U &&>)
-        : m_has_value(true)
+        : m_null(), m_has_value(true)
     {
-        new (&m_storage) T(forward<U>(value));
+        construct_at(&m_storage, forward<U>(value));
```

Comment:
Couldnt this directly construct `m_storage` now?

Reply:
Do you mean in the constructor's initializer list? I've tried to change it into
```c++
        : m_storage(move(value)), m_has_value(true)
    {
    }
```
and that fails with
```
[197/3247] Generating LocaleData.h, LocaleData.cpp
FAILED: Userland/Libraries/LibLocale/LocaleData.h Userland/Libraries/LibLocale/LocaleData.cpp /home/nick/src/serenity/Build/lagom/Userland/Libraries/LibLocale/LocaleData.h /home/nick/src/serenity/Build/lagom/Userland/Libraries/LibLocale/LocaleData.cpp 
cd /home/nick/src/serenity/Build/lagom/Userland/Libraries/LibLocale && /home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData -h LocaleData.h.tmp -c LocaleData.cpp.tmp -b /home/nick/src/serenity/Build/superbuild-x86_64/../caches/CLDR/cldr-bcp47 -r /home/nick/src/serenity/Build/superbuild-x86_64/../caches/CLDR/cldr-core -l /home/nick/src/serenity/Build/superbuild-x86_64/../caches/CLDR/cldr-localenames-modern -m /home/nick/src/serenity/Build/superbuild-x86_64/../caches/CLDR/cldr-misc-modern -n /home/nick/src/serenity/Build/superbuild-x86_64/../caches/CLDR/cldr-numbers-modern -d /home/nick/src/serenity/Build/superbuild-x86_64/../caches/CLDR/cldr-dates-modern && /nix/store/a9ny340wpz79v8lk059c10gh4dqbll94-cmake-3.29.2/bin/cmake -E copy_if_different LocaleData.h.tmp LocaleData.h && /nix/store/a9ny340wpz79v8lk059c10gh4dqbll94-cmake-3.29.2/bin/cmake -E copy_if_different LocaleData.cpp.tmp LocaleData.cpp && /nix/store/a9ny340wpz79v8lk059c10gh4dqbll94-cmake-3.29.2/bin/cmake -E remove LocaleData.h.tmp LocaleData.cpp.tmp
VERIFICATION FAILED: m_has_value at /home/nick/src/serenity/Meta/Lagom/../../AK/Optional.h:212
/home/nick/src/serenity/Build/lagom/bin/../lib64/liblagom-ak.so.0(ak_verification_failed+0xaa) [0x7f272832267a]
/home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData() [0x225f5c]
/home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData() [0x226f0d]
/home/nick/src/serenity/Build/lagom/bin/../lib64/liblagom-coreminimal.so.0 Core::Directory::for_each_entry(Core::DirIterator::Flags, AK::Function<AK::ErrorOr<AK::IterationDecision, AK::Error> (Core::DirectoryEntry const&, Core::Directory const&)>) 0x114) [0x7f2728388fe4]
/home/nick/src/serenity/Build/lagom/bin/../lib64/liblagom-coreminimal.so.0 Core::Directory::for_each_entry(AK::StringView, Core::DirIterator::Flags, AK::Function<AK::ErrorOr<AK::IterationDecision, AK::Error> (Core::DirectoryEntry const&, Core::Directory const&)>) 0x1e6) [0x7f2728389466]
/home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData() [0x21f8ca]
/home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData() [0x221498]
/home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData() [0x23bfb8]
/nix/store/dbcw19dshdwnxdv5q2g6wldj6syyvq7l-glibc-2.39-52/lib/libc.so.6(+0x2a10e) [0x7f2727d5a10e]
/nix/store/dbcw19dshdwnxdv5q2g6wldj6syyvq7l-glibc-2.39-52/lib/libc.so.6(__libc_start_main+0x89) [0x7f2727d5a1c9]
/home/nick/src/serenity/Build/lagom/bin/GenerateLocaleData() [0x20fb05]
```
I don't understand why it fails yet, but I'll try to find out

Reply:
`forward` instead of `move` is likely a contributing factor

Reply:
Good point, I somehow missed that `value` is a forwarding reference there.